File: ios/Classes/Convert.swift
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import Foundation
import GoogleMaps
import GoogleNavigation
import GoogleRidesharingDriver

enum Convert {
  // This conversion functions has been duplicated from
  // google_navigation_flutter. Keep in sync.
  static func convertLatLngFromDto(point: LatLngDto) -> CLLocationCoordinate2D {
    CLLocationCoordinate2D(
      latitude: point.latitude,
      longitude: point.longitude
    )
  }

  // This conversion functions has been duplicated from
  // google_navigation_flutter. Keep in sync.
  static func convertLatLngToDto(point: CLLocationCoordinate2D) -> LatLngDto {
    LatLngDto(latitude: point.latitude, longitude: point.longitude)
  }

  // This conversion functions has been duplicated from
  // google_navigation_flutter. Keep in sync.
  static func convertNavigationWayPointToDto(_ gmsNavigationWaypoint: GMSNavigationWaypoint?)
    -> NavigationWaypointDto? {
    guard let gmsNavigationWaypoint else { return nil }
    return NavigationWaypointDto(
      title: gmsNavigationWaypoint.title,
      target: .init(
        latitude: gmsNavigationWaypoint.coordinate.latitude,
        longitude: gmsNavigationWaypoint.coordinate.longitude
      ),
      placeID: gmsNavigationWaypoint.placeID,
      preferSameSideOfRoad: gmsNavigationWaypoint.preferSameSideOfRoad,
      preferredSegmentHeading: Int64(gmsNavigationWaypoint.preferredHeading)
    )
  }

  // This conversion functions has been duplicated from
  // google_navigation_flutter. Keep in sync.
  static func convertNavigationWayPointFromDto(_ waypoint: NavigationWaypointDto?)
    -> GMSNavigationWaypoint? {
    guard let waypoint else { return nil }
    if let latitude = waypoint.target?.latitude, let longitude = waypoint.target?.longitude {
      if let preferSameSideOfRoad = waypoint.preferSameSideOfRoad {
        return GMSNavigationWaypoint(
          location: .init(latitude: latitude, longitude: longitude),
          title: waypoint.title,
          preferSameSideOfRoad: preferSameSideOfRoad
        )
      } else if let preferredSegmentHeading = waypoint.preferredSegmentHeading {
        return GMSNavigationWaypoint(
          location: .init(latitude: latitude, longitude: longitude),
          title: waypoint.title,
          // TODO: Handle the 32bit x 64bit conversion correctly.
          preferredSegmentHeading: Int32(preferredSegmentHeading)
        )
      }
      return GMSNavigationWaypoint(
        location: CLLocationCoordinate2D(
          latitude: latitude,
          longitude: longitude
        ),
        title: waypoint.title
      )
    }
    if let placeID = waypoint.placeID {
      return GMSNavigationWaypoint(
        placeID: placeID,
        title: waypoint.title
      )
    }
    return nil
  }

  // This conversion functions has been duplicated from
  // google_navigation_flutter. Keep in sync.
  static func convertWaypoints(_ waypoints: [NavigationWaypointDto?])
    -> [GMSNavigationWaypoint] {
    waypoints
      .map { waypoint -> GMSNavigationWaypoint? in
        guard let waypoint else { return nil }
        return convertNavigationWayPointFromDto(waypoint)
      }
      .compactMap { $0 }
  }

  // Converts Pigeon [TaskInfoDto] to Google Driver [GMTSTaskInfo].
  static func convertTaskInfoFromDto(task: TaskInfoDto) -> GMTSTaskInfo {
    GMTSTaskInfo(taskID: task.taskId, taskDuration: TimeInterval(task.durationSeconds))
  }

  // Converts Google Driver [GMTSTaskInfo] to Pigeon [TaskInfoDto].
  static func convertTaskInfoToDto(task: GMTSTaskInfo) -> TaskInfoDto {
    TaskInfoDto(taskId: task.taskID, durationSeconds: Int64(task.taskDuration))
  }

  // Converts pigeon [VehicleStopStateDto] to Google Drive [Int].
  static func convertVehicleStopStateFromDto(vehicleStopState: VehicleStopStateDto)
    -> GMTDVehicleStopState {
    switch vehicleStopState {
    case VehicleStopStateDto.stateUnspecified:
      return .unspecified
    case VehicleStopStateDto.newStop:
      return .new
    case VehicleStopStateDto.enroute:
      return .enroute
    case VehicleStopStateDto.arrived:
      return .arrived
    }
  }

  // Converts Google Drive [Int] to pigeon [VehicleStopStateDto].
  static func convertVehicleStopStateToDto(googleVehicleStopState: GMTDVehicleStopState)
    -> VehicleStopStateDto {
    switch googleVehicleStopState {
    case .unspecified:
      return VehicleStopStateDto.stateUnspecified
    case .new:
      return VehicleStopStateDto.newStop
    case .enroute:
      return VehicleStopStateDto.enroute
    case .arrived:
      return VehicleStopStateDto.arrived
    @unknown default:
      return VehicleStopStateDto.stateUnspecified
    }
  }

  // Convert GMTDVehicleStop into VehicleStopDto
  static func convertVehicleStopToDto(stop: GMTDVehicleStop) -> VehicleStopDto {
    VehicleStopDto(
      vehicleStopState: convertVehicleStopStateToDto(googleVehicleStopState: stop.state),
      waypoint: convertNavigationWayPointToDto(stop.plannedWaypoint),
      taskInfoList: stop.taskInfoArray.map { taskInfo -> TaskInfoDto? in
        convertTaskInfoToDto(task: taskInfo)
      }
    )
  }

  // Converts Google Driver [GMTDVehicleStop] to Pigeon [VehicleStopDto].
  static func convertVehicleStopFromDto(stop: VehicleStopDto) -> GMTDVehicleStop {
    GMTDVehicleStop(taskInfoArray: stop.taskInfoList.map { taskInfo -> GMTSTaskInfo? in
      guard let taskInfo else { return nil }

      return convertTaskInfoFromDto(task: taskInfo)
    }.compactMap { $0 }, plannedWaypoint: convertNavigationWayPointFromDto(stop.waypoint),
    state: convertVehicleStopStateFromDto(
      vehicleStopState: stop.vehicleStopState
    ))
  }

  // Converts Pigeon [VehicleStateDto] to Google Driver [GMTDVehicleState].
  static func convertVehicleStateFromDto(state: VehicleStateDto)
    -> GMTDVehicleState {
    switch state {
    case .offline:
      return GMTDVehicleState.offline
    case .online:
      return GMTDVehicleState.online
    }
  }

  // Converts Google Driver [GMTDVehicleState] to Pigeon [VehicleStateDto].
  static func convertVehicleStateToDto(state: GMTDVehicleState)
    -> VehicleStateDto {
    switch state {
    case .offline:
      return VehicleStateDto.offline
    case .online:
      return VehicleStateDto.online
    @unknown default:
      // Should not happen.
      return VehicleStateDto.offline
    }
  }

  static func convertVehicleUpdateToDto(vehicleUpdate: GMTDVehicleUpdate,
                                        ridesharing: Bool) -> VehicleUpdateDto {
    VehicleUpdateDto(
      vehicleState: ridesharing ?
        convertVehicleStateToDto(state: vehicleUpdate.vehicleState)
        : nil,
      location: vehicleUpdate
        .location != nil ? convertLatLngToDto(point: vehicleUpdate.location!.coordinate) : nil,
      destinationWaypoint: convertNavigationWayPointToDto(vehicleUpdate.destinationWaypoint),
      route: (vehicleUpdate.route ?? []).map { convertLatLngToDto(point: $0.coordinate)
      },
      remainingTimeInSeconds: vehicleUpdate.remainingTimeInSeconds?.doubleValue,
      remainingDistanceInMeters: vehicleUpdate.remainingDistanceInMeters?.doubleValue
    )
  }

  static func convertGRPCErrorCodes(errorCode: Int) -> String {
    let errorCodes = [
      0: "OK",
      1: "CANCELLED",
      2: "UNKNOWN",
      3: "INVALID_ARGUMENT",
      4: "DEADLINE_EXCEEDED",
      5: "NOT_FOUND",
      6: "ALREADY_EXISTS",
      7: "PERMISSION_DENIED",
      8: "RESOURCE_EXHAUSTED",
      9: "FAILED_PRECONDITION",
      10: "ABORTED",
      11: "OUT_OF_RANGE",
      12: "UNIMPLEMENTED",
      14: "UNAVAILABLE",
      15: "DATA_LOSS",
      16: "UNAUTHENTICATED",
    ]
    return errorCodes[errorCode] ?? "UNKNOWN"
  }

  static func convertDeliveryVehicleToDto(deliveryVehicle: GMTDDeliveryVehicle)
    -> DeliveryVehicleDto {
    DeliveryVehicleDto(
      providerId: deliveryVehicle.providerID,
      id: deliveryVehicle.vehicleID,
      name: deliveryVehicle.vehicleName,
      stops: deliveryVehicle.vehicleStops?.map { stop in
        convertVehicleStopToDto(stop: stop)
      } ?? []
    )
  }
}

File: android/src/main/kotlin/com/google/maps/flutter/driver/Convert.kt
/*
 * Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.maps.flutter.driver

import android.location.Location
import android.os.Build
import com.google.android.gms.maps.model.LatLng
import com.google.android.libraries.mapsplatform.transportation.driver.api.base.data.DriverContext
import com.google.android.libraries.mapsplatform.transportation.driver.api.base.data.TaskInfo
import com.google.android.libraries.mapsplatform.transportation.driver.api.base.data.VehicleStop
import com.google.android.libraries.mapsplatform.transportation.driver.api.delivery.data.DeliveryVehicle
import com.google.android.libraries.mapsplatform.transportation.driver.api.ridesharing.vehiclereporter.RidesharingVehicleReporter
import com.google.android.libraries.navigation.Waypoint
import com.google.common.collect.Lists

/** Converters from and to Pigeon generated values. */
object Convert {
  /**
   * Converts Pigeon [LatLngDto] to Google Maps [LatLng].
   *
   * This function has been duplicated from google_navigation_flutter package. Keep in sync.
   *
   * @param point Pigeon [LatLngDto].
   * @return Google Maps [LatLng].
   */
  fun convertLatLngFromDto(point: LatLngDto): LatLng {
    return LatLng(point.latitude, point.longitude)
  }

  /**
   * Converts Google Maps [LatLng] to Pigeon [LatLngDto].
   *
   * This function has been duplicated from google_navigation_flutter package. Keep in sync.
   *
   * @param point Google Maps [LatLng].
   * @return Pigeon [LatLngDto].
   */
  fun convertLatLngToDto(point: LatLng): LatLngDto {
    return LatLngDto(point.latitude, point.longitude)
  }

  /**
   * Converts pigeon [NavigationWaypointDto] to Google Navigation [Waypoint].
   *
   * This function has been duplicated from google_navigation_flutter package. Keep in sync.
   *
   * @param waypoint pigeon [NavigationWaypointDto].
   * @return Google Navigation [Waypoint].
   */
  fun convertWaypointFromDto(waypoint: NavigationWaypointDto): Waypoint {
    val builder = Waypoint.builder()
    if (waypoint.target != null) {
      builder.setLatLng(waypoint.target.latitude, waypoint.target.longitude)
    }
    if (waypoint.preferSameSideOfRoad == true) {
      builder.setPreferSameSideOfRoad(true)
    }
    if (waypoint.preferredSegmentHeading != null) {
      builder.setPreferredHeading(waypoint.preferredSegmentHeading.toInt())
    }
    if (waypoint.placeID != null) {
      builder.setPlaceIdString(waypoint.placeID)
    }
    builder.setTitle(waypoint.title)
    return builder.build()
  }

  /**
   * Converts Google Navigation [Waypoint] to pigeon [NavigationWaypointDto].
   *
   * This function has been duplicated from google_navigation_flutter package. Keep in sync.
   *
   * @param waypoint Google Navigation [Waypoint].
   * @return pigeon [NavigationWaypointDto].
   */
  private fun convertWaypointToDto(waypoint: Waypoint): NavigationWaypointDto {
    return NavigationWaypointDto(
      waypoint.title,
      convertLatLngToDto(waypoint.position),
      waypoint.placeId,
      waypoint.preferSameSideOfRoad,
      waypoint.preferredHeading.takeIf { it != -1 }?.toLong(),
    )
  }

  /**
   * Converts Pigeon [TaskInfoDto] to Google Driver [TaskInfo].
   *
   * @param task Pigeon [TaskInfoDto].
   * @return Google Driver [TaskInfo].
   */
  fun convertTaskInfoFromDto(task: TaskInfoDto): TaskInfo {
    return TaskInfo.builder()
      .setTaskId(task.taskId)
      .setTaskDurationSeconds(task.durationSeconds)
      .build()
  }

  /**
   * Converts Google Driver [TaskInfo] to Pigeon [TaskInfoDto].
   *
   * @param task Google Driver [TaskInfo].
   * @return Pigeon [TaskInfoDto].
   */
  fun convertTaskInfoToDto(task: TaskInfo): TaskInfoDto {
    return TaskInfoDto(task.taskId, task.taskDurationSeconds)
  }

  /**
   * Converts pigeon [VehicleStopStateDto] to Google Drive [Int].
   *
   * @param vehicleStopState pigeon [VehicleStopStateDto].
   * @return VehicleStop.VehicleStopState [Int].
   */
  fun convertVehicleStopStateFromDto(vehicleStopState: VehicleStopStateDto): Int {
    return when (vehicleStopState) {
      VehicleStopStateDto.STATE_UNSPECIFIED -> VehicleStop.VehicleStopState.UNSPECIFIED
      VehicleStopStateDto.NEW_STOP -> VehicleStop.VehicleStopState.NEW
      VehicleStopStateDto.ENROUTE -> VehicleStop.VehicleStopState.ENROUTE
      VehicleStopStateDto.ARRIVED -> VehicleStop.VehicleStopState.ARRIVED
    }
  }

  /**
   * Converts Google Drive [Int] to pigeon [VehicleStopStateDto].
   *
   * @param googleVehicleStopState [Int].
   * @return pigeon [VehicleStopStateDto].
   */
  fun convertVehicleStopStateToDto(googleVehicleStopState: Int): VehicleStopStateDto {
    return when (googleVehicleStopState) {
      VehicleStop.VehicleStopState.UNSPECIFIED -> VehicleStopStateDto.STATE_UNSPECIFIED
      VehicleStop.VehicleStopState.NEW -> VehicleStopStateDto.NEW_STOP
      VehicleStop.VehicleStopState.ENROUTE -> VehicleStopStateDto.ENROUTE
      VehicleStop.VehicleStopState.ARRIVED -> VehicleStopStateDto.ARRIVED
      else -> {
        VehicleStopStateDto.STATE_UNSPECIFIED
      }
    }
  }

  /**
   * Converts Google Driver [VehicleStop] to Pigeon [VehicleStopDto].
   *
   * @param stop Google Driver [VehicleStop].
   * @return Pigeon [VehicleStopDto].
   */
  fun convertVehicleStopToDto(stop: VehicleStop): VehicleStopDto {
    return VehicleStopDto(
      convertVehicleStopStateToDto(stop.vehicleStopState),
      convertWaypointToDto(stop.waypoint),
      Lists.transform<TaskInfo, TaskInfoDto?>(stop.taskInfoList) { task: TaskInfo ->
        convertTaskInfoToDto(task)
      },
    )
  }

  /**
   * Converts Google Driver [VehicleStop] to Pigeon [VehicleStopDto].
   *
   * @param stop Google Driver [VehicleStop].
   * @return Pigeon [VehicleStopDto].
   */
  fun convertVehicleStopFromDto(stop: VehicleStopDto): VehicleStop {
    val builder: VehicleStop.Builder =
      VehicleStop.builder()
        .setVehicleStopState(convertVehicleStopStateFromDto(stop.vehicleStopState))
    if (stop.waypoint != null) {
      builder.setWaypoint(convertWaypointFromDto(stop.waypoint))
    }

    builder.setTaskInfoList(
      Lists.transform<TaskInfoDto, TaskInfo>(stop.taskInfoList.filterNotNull()) { task: TaskInfoDto
        ->
        convertTaskInfoFromDto(task)
      }
    )
    return builder.build()
  }

  /**
   * Converts pigeon [VehicleStateDto] to Google Drive [RidesharingVehicleReporter.VehicleState].
   *
   * @param state pigeon [VehicleStateDto].
   * @return [RidesharingVehicleReporter.VehicleState].
   */
  fun convertVehicleStateFromDto(state: VehicleStateDto): RidesharingVehicleReporter.VehicleState {
    return when (state) {
      VehicleStateDto.OFFLINE -> RidesharingVehicleReporter.VehicleState.OFFLINE
      VehicleStateDto.ONLINE -> RidesharingVehicleReporter.VehicleState.ONLINE
    }
  }

  fun extractErrorCode(t: Throwable): String? {
    val messages = t.message?.split(":", limit = 2)
    return messages?.getOrNull(0)?.trimStart()
  }

  fun extractErrorMessage(t: Throwable): String? {
    val messages = t.message?.split(":", limit = 2)
    return messages?.getOrNull(1)?.trimStart()
  }

  fun convertToDriverException(t: Throwable): FlutterError {
    when (t) {
      is java.util.concurrent.ExecutionException -> {
        return FlutterError(
          "driverException",
          t.message?.replace("java.lang.RuntimeException: Exception: ", "")
            ?: "Token retrieval from the backend failed.",
          "UNAUTHENTICATED",
        )
      }
      else -> {
        return FlutterError(
          "driverException",
          extractErrorMessage(t) ?: "Driver API call failed.",
          extractErrorCode(t) ?: "UNKNOWN",
        )
      }
    }
  }

  /**
   * Converts pigeon [LocationDto] to Android [Location].
   *
   * @param location pigeon [LocationDto].
   * @return [Location].
   */
  fun convertLocationFromDto(location: LocationDto): Location {
    return Location(location.provider).apply {
      if (location.accuracy != null) {
        accuracy = location.accuracy.toFloat()
      }
      if (location.altitude != null) {
        altitude = location.altitude
      }
      if (location.elapsedRealtimeNanos != null) {
        elapsedRealtimeNanos = location.elapsedRealtimeNanos
      }
      if (location.bearing != null) {
        bearing = location.bearing.toFloat()
      }
      if (location.isMock != null && Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
        isMock = location.isMock
      }
      if (location.latitude != null) {
        latitude = location.latitude
      }
      if (location.longitude != null) {
        longitude = location.longitude
      }
      if (location.speed != null) {
        speed = location.speed.toFloat()
      }
      if (location.time != null) {
        time = location.time
      }
    }
  }

  /**
   * Converts Android [Location] to pigeon [LocationDto].
   *
   * @param location Android [Location].
   * @return [LocationDto].
   */
  fun convertLocationToDto(location: Location): LocationDto {
    val isMock =
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
        location.isMock
      } else {
        null
      }
    return LocationDto(
      provider = location.provider,
      accuracy = location.accuracy.toDouble(),
      altitude = location.altitude,
      elapsedRealtimeNanos = location.elapsedRealtimeNanos,
      bearing = location.bearing.toDouble(),
      latitude = location.latitude,
      longitude = location.longitude,
      isMock = isMock,
      speed = location.speed.toDouble(),
      time = location.time,
    )
  }

  /**
   * Converts Google Driver [DeliveryVehicle] to pigeon [DeliveryVehicleDto].
   *
   * @param deliveryVehicle Google Driver [DeliveryVehicleDto].
   * @return pigeon [DeliveryVehicleDto].
   */
  fun convertDeliveryVehicleToDto(deliveryVehicle: DeliveryVehicle): DeliveryVehicleDto {
    return DeliveryVehicleDto(
      providerId = deliveryVehicle.providerId,
      id = deliveryVehicle.vehicleId,
      name = deliveryVehicle.vehicleName,
      stops = deliveryVehicle.vehicleStops.map { convertVehicleStopToDto(it) },
    )
  }

  /**
   * Converts Google Drive [DriverContext.DriverStatusListener.StatusLevel] to pigeon
   * [DriverStatusLevelDto].
   *
   * @param statusLevel [DriverContext.DriverStatusListener.StatusLevel].
   * @return pigeon [DriverStatusLevelDto].
   */
  fun convertStatusLevelToDto(
    statusLevel: DriverContext.DriverStatusListener.StatusLevel
  ): DriverStatusLevelDto {
    return when (statusLevel) {
      DriverContext.DriverStatusListener.StatusLevel.DEBUG -> DriverStatusLevelDto.DEBUG
      DriverContext.DriverStatusListener.StatusLevel.INFO -> DriverStatusLevelDto.INFO
      DriverContext.DriverStatusListener.StatusLevel.WARNING -> DriverStatusLevelDto.WARNING
      DriverContext.DriverStatusListener.StatusLevel.ERROR -> DriverStatusLevelDto.ERROR
    }
  }

  /**
   * Converts Google Drive [DriverContext.DriverStatusListener.StatusCode] to pigeon
   * [DriverStatusCodeDto].
   *
   * @param statusCode [DriverContext.DriverStatusListener.StatusCode].
   * @return pigeon [DriverStatusCodeDto].
   */
  fun convertStatusCodeToDto(
    statusCode: DriverContext.DriverStatusListener.StatusCode
  ): DriverStatusCodeDto {
    return when (statusCode) {
      DriverContext.DriverStatusListener.StatusCode.DEFAULT -> DriverStatusCodeDto.DEFAULT_STATUS
      DriverContext.DriverStatusListener.StatusCode.UNKNOWN_ERROR ->
        DriverStatusCodeDto.UNKNOWN_ERROR
      DriverContext.DriverStatusListener.StatusCode.VEHICLE_NOT_FOUND ->
        DriverStatusCodeDto.VEHICLE_NOT_FOUND
      DriverContext.DriverStatusListener.StatusCode.BACKEND_CONNECTIVITY_ERROR ->
        DriverStatusCodeDto.BACKEND_CONNECTIVITY_ERROR
      DriverContext.DriverStatusListener.StatusCode.PERMISSION_DENIED ->
        DriverStatusCodeDto.PERMISSION_DENIED
      DriverContext.DriverStatusListener.StatusCode.SERVICE_ERROR ->
        DriverStatusCodeDto.SERVICE_ERROR
      DriverContext.DriverStatusListener.StatusCode.FILE_ACCESS_ERROR ->
        DriverStatusCodeDto.FILE_ACCESS_ERROR
      DriverContext.DriverStatusListener.StatusCode.TRAVELED_ROUTE_ERROR ->
        DriverStatusCodeDto.TRAVELED_ROUTE_ERROR
    }
  }
}
